serverless-heaven/serverless-webpack
# fix(packaging): ensure consistent artifact sha

## What did you implement:

Closes #881

bestzip does not support sorting contents or stripping file creation date or modified date, resulting in an indeterminate archive. This breaks serverless' change detection as the zip file created by serverless-webpack has a different checksum each time it is generated, even if the source did not change.



## How did you implement it:

Copied over serverless' zip method as it cannot be used when imported by plugin (the piece we need is coupled with other functionality we need to bypass).

Disabled the native zip method and bypass bestzip by default, as bestzip is not configurable and does not guarantee ordering of contents inside of zip file.

Added new configuration "useNativeZip" if you still want the old behavior, but use the serverless method instead of bestzip's "nodezip" as fallback.



## How can we verify it:



Examples:
* With `useNativeZip: true` in webpack config of serverless.yml, run `serverless package` twice and compare checksums. You will see they have changed. Alternatively, run `serverless deploy` and you will see that serverless detected a change. 
* Now remove that config to use the new default behavior and repeat - you will get the same checksum each time.


## Todos:

- [] Enable "Allow edits from maintainers" for this PR

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
